pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -3543,98 +3543,90 @@ protected void internalExpireMessagesByPosition(AsyncResponse asyncResponse, Str
+                                } else if (th instanceof WebApplicationException) {
+                                    asyncResponse.resume(th);
+                                    return null;
+                                } else {
+                                    log.error("[{}] Failed to trigger compaction on topic {}",
+                                            clientAppId(), topicName, exception);
+                                    asyncResponse.resume(new RestException(exception));
+                                    return null;
+                                }
                             }
+                            asyncResponse.resume(Response.noContent().build());
+                            return null;
+                        });
+                    } else {
+                        try {
+                            internalTriggerCompactionNonPartitionedTopic(authoritative);
+                        } catch (Exception e) {
+                            if (e instanceof WebApplicationException) {
+                                if (log.isDebugEnabled()) {
+                                    log.debug("[{}] Failed to trigger compaction on topic {}, "
+                                            + "redirecting to other brokers.", clientAppId(), topicName, e);
+                                }
+                                resumeAsyncResponseExceptionally(asyncResponse, e);
+                                return;
+                            }
+                            log.error("[{}] Failed to trigger compaction on topic {}", clientAppId(), topicName, e);
                             resumeAsyncResponseExceptionally(asyncResponse, e);
                             return;
                         }
-                        log.error("[{}] Failed to trigger compaction on topic {}", clientAppId(), topicName, e);
-                        resumeAsyncResponseExceptionally(asyncResponse, e);
-                        return;
+                        asyncResponse.resume(Response.noContent().build());
                     }
-                    asyncResponse.resume(Response.noContent().build());
-                }
-            }).exceptionally(ex -> {
-                log.error("[{}] Failed to trigger compaction on topic {}", clientAppId(), topicName, ex);
-                resumeAsyncResponseExceptionally(asyncResponse, ex);
-                return null;
-            });
-        }
+                }).exceptionally(ex -> {
+                    Throwable cause = ex.getCause();
+                    log.error("[{}] Failed to validate global namespace ownership to trigger compaction on topic {}",
+                            clientAppId(), topicName, cause);
+                    resumeAsyncResponseExceptionally(asyncResponse, cause);
+                    return null;
+                });
+            }
+        });
     }
 
     protected void internalTriggerCompactionNonPartitionedTopic(boolean authoritative) {
```

Comment:
This method also has the problem of asynchronous methods call synchronous methods.

@liudezhi2098  
We need to check ``internalTriggerCompaction`` and the methods it calls internally :-)

Reply:
missed one，PTAL:-)